freeipa/src/main/java/com/sequenceiq/freeipa/sync/FreeipaChecker.java

```diff
@@ -29,31 +32,31 @@
     @Inject
     private FreeIpaHealthDetailsService freeIpaHealthDetailsService;
 
-    private Pair<List<DetailedStackStatus>, String> checkStatus(Stack stack, Set<InstanceMetaData> checkableInstances) throws Exception {
+    private Pair<Map<InstanceMetaData, DetailedStackStatus>, String> checkStatus(Stack stack, Set<InstanceMetaData> checkableInstances) throws Exception {
         return checkedMeasure(() -> {
-            List<DetailedStackStatus> statuses = new LinkedList<>();
+            Map<InstanceMetaData, DetailedStackStatus> statuses = new HashMap<>();
             List<RPCResponse<Boolean>> responses = new LinkedList<>();
             for (InstanceMetaData instanceMetaData : checkableInstances) {
                 try {
                     RPCResponse<Boolean> response = checkedMeasure(() -> freeIpaHealthDetailsService.checkFreeIpaHealth(stack, instanceMetaData), LOGGER,
                             ":::Auto sync::: FreeIPA health check ran in {}ms");
                     responses.add(response);
                     if (response.getResult()) {
-                        statuses.add(DetailedStackStatus.AVAILABLE);
+                        statuses.put(instanceMetaData, DetailedStackStatus.AVAILABLE);
                     } else {
-                        statuses.add(DetailedStackStatus.UNHEALTHY);
+                        statuses.put(instanceMetaData, DetailedStackStatus.UNHEALTHY);
                     }
                 } catch (Exception e) {
                     LOGGER.info("FreeIpaClientException occurred during status fetch: " + e.getMessage(), e);
-                    statuses.add(DetailedStackStatus.UNREACHABLE);
+                    statuses.put(instanceMetaData, DetailedStackStatus.UNREACHABLE);
                 }
             }
             String message = getMessages(responses);
             return Pair.of(statuses, message);
         }, LOGGER, ":::Auto sync::: freeipa server status is checked in {}ms");
     }
 
-    public SyncResult getStatus(Stack stack, Set<InstanceMetaData> checkableInstances) {
+    public SyncResult getStatus(Stack stack, Set<InstanceMetaData> checkableInstances, Map<InstanceMetaData, DetailedStackStatus> instanceHealthStatusMap) {
```

Comment:
Instead of using an outbound parameter `instanceHealthStatusMap`, would it be better to include that as a part of the `SyncResult`?

Reply:
This is great suggestion. I returned instanceHealthStatusMap in SyncResult now. It will make the code less coupled and more readable. I removed the carrier parameter.